### ISSUE TITLE
improper_ctypes lint doesn't detect cross-crate un-ffi-safe structs

### ISSUE THREAD

**Issue by tomjakubowski** [rust-lang/rust#20098]
For example, it accepts this:

```rust
#![crate_type="lib"]
#![deny(improper_ctypes)]

extern "C" {
    pub fn foo(x: String);
}
```

Even though `String` is not FFI safe.

**Comment by tomjakubowski** [comment]
This and #19834 are still serious and might be masking bugs in programs exporting a native interface. I'll see if I can revive my PR from a few months ago.

**Comment by TimNN** [comment]
This seems to be fixed in the current stable: http://is.gd/Ew2jCI

**Comment by alexcrichton** [comment]
Yay, thanks @TimNN!
